### PR TITLE
HttpURL parsing.

### PR DESCRIPTION
**Pull request by swankjesse**
This includes everything but character encoding. The implementation
is built around one oversized method that runs top to bottom; I found
this was easier to follow than the switch statement recommended by the
URL standard.

### CODE REVIEW

File: okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java [hunk 1]
```diff
@@ -32,70 +32,107 @@
...
   }
 
   /**
-   * Returns the decoded (potentially non-ASCII) hostname. The returned string may contain non-ASCII
-   * characters and is <strong>not suitable</strong> for DNS lookups; for that use {@link
-   * #encodedHost}. For example, this may return {@code ☃.net} which is a user-displayable IDN that
-   * cannot be used for DNS lookups without encoding.
+   * Returns the host address suitable for use with {@link InetAddress#getAllByName(String)}. May
+   * be a regular host name ({@code android.com}), an IPv4 address ({@code 127.0.0.1}), an IPv6
+   * address ({@code ::1}; note that there are no square braces), or an encoded IDN ({@code
```
Comment (JakeWharton) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: missing closing parenthesis on IPv6 example

Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: Fixed.

File: okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java [hunk 2]
```diff
@@ -228,75 +281,412 @@ public Builder encodedPassword(String encodedPassword) {
...
+
+      return new HttpUrl(scheme, username, effectivePassword, host, effectivePort, effectivePath,
+          query, fragment, url.toString());
+    }
+
+    HttpUrl parse(HttpUrl base, String input) {
+      int pos = 0;
+      int limit = input.length();
+      pos += leadingAsciiWhitespace(input, pos, limit);
+      limit = trailingAsciiWhitespace(input, pos, limit);
```
Comment (JakeWharton) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: Both of these methods are one-shots. Why not change behavior so you can do:
```java
int pos = skipLeadingAsciiWhitespace(input);
int limit = skipTrailingAsciiWhitespace(input, pos);
```

Comment (JakeWharton) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: The only reason I commented was the non-symmetric `

Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: Oooh, I like that better.

**Comment by JakeWharton** [comment]
LGTM

**Comment by swankjesse** [comment]
Working towards https://github.com/square/okhttp/issues/1486

**Pull request merged**